### PR TITLE
Transparently Handle Object/Array object types von get/set

### LINKED ISSUE #830: Handling of objets in datapoints has to changed, because new js-controller only allows JSON
[Änderungen im js-controller](https://github.com/ioBroker/ioBroker.js-controller/pull/1319) machen es erforderlich, die Behandlung von Datenpunkten vom Typ "object" und "array" zu ändern.

### PR DESCRIPTION
**Pull request by Apollon77**
# fixes #830

### CODE REVIEW

**Comment by Apollon77** [comment]
@paul53 Please also have a look ... makes that sense? Do I covered all cases?

**Comment by Apollon77** [comment]
We also need to define if this is "breaking"!? So Major release or minor?

**Comment by GermanBluefox** [comment]
Should be OK. But at least minor step in versioning must be done.

File: lib/sandbox.js [hunk 1]
```diff
@@ -277,6 +277,24 @@ function sandBox(script, name, verbose, debug, context) {
                     + `This warning might become an error in future versions.`
                 );
             }
+            if (!isBinary && (actualCommonType === 'array' || actualCommonType === 'object')) {
+                try {
+                    if (state && isObject(state) && typeof state.val !== 'undefined') {
```
Comment (AlCalzone) [inline lib/sandbox.js:282]:
```suggestion
                    if (isObject(state) && typeof state.val !== 'undefined') {
```
`isObject` returns `false` for `null` and `undefined`

File: lib/sandbox.js [hunk 2]
```diff
@@ -1618,9 +1630,9 @@ function sandBox(script, name, verbose, debug, context) {
         getState:       function (id, callback) {
             if (typeof callback === 'function') {
                 if (id.indexOf('.') === -1) {
-                    adapter.getState(id, callback);
+                    adapter.getState(id, (err, state) => callback(err, context.convertBackStringifiedValues(id, state)));
```
Comment (AlCalzone) [inline lib/sandbox.js:1633]: If this is going to be used more often, this callback wrapper should be refactored into a separate method.

File: main.js [hunk 3]
```diff
@@ -253,6 +254,21 @@ const regExGlobalNew = /script\.js\.global\./;
 function checkIsGlobal(obj) {
     return obj && obj.common && (regExGlobalOld.test(obj.common.name) || regExGlobalNew.test(obj._id));
 }
+
+function convertBackStringifiedValues(id, state) {
+    if (state && state.val && typeof state.val === 'string' &&
```
Comment (AlCalzone) [inline main.js:259]:
```suggestion
    if (state && typeof state.val === 'string' &&
```
not necessary.

File: main.js [hunk 4]
```diff
@@ -253,6 +254,21 @@ const regExGlobalNew = /script\.js\.global\./;
...
+        context.objects[id] && context.objects[id].common && context.objects[id].common.type &&
+        (context.objects[id].common.type === 'array' || context.objects[id].common.type === 'object')) {
```
Comment (AlCalzone) [inline main.js:261]:
```suggestion
        context.objects[id] && context.objects[id].common &&
        (context.objects[id].common.type === 'array' || context.objects[id].common.type === 'object')) {
```
Also not necessary, it gets checked afterwards.

**Review by AlCalzone — CHANGES_REQUESTED** [review]
just a few nits

**Comment by AlCalzone** [comment]
> "breaking"

Isn't that rather a fix?

File: lib/sandbox.js (hunk 1 above)
Comment (Apollon77) [reply lib/sandbox.js:282]: Right

File: lib/sandbox.js (hunk 2 above)
Comment (Apollon77) [reply lib/sandbox.js:1633]: only those some places ... would leave for now

Comment (Apollon77) [reply lib/sandbox.js:1633]: PS: But yes I also thought that

File: main.js (hunk 3 above)
Comment (Apollon77) [reply main.js:259]: true

**Comment by Apollon77** [comment]
> Isn't that rather a fix?

yes, but in fact on data level the state content is different ....no idea if this can have issues in "interoperability"

**Comment by Apollon77** [comment]
Comments included

**Pull request merged**